src/core_plugins/metrics/public/components/data_config.js

```diff
@@ -0,0 +1,102 @@
+import React from 'react';
+import createTextHandler from './lib/create_text_handler';
+import createSelectHandler from './lib/create_select_handler';
+import Select from 'react-select';
+import { Code } from './code';
+import { Note } from './note';
+import { Info } from './info';
+import { htmlIdGenerator } from '@elastic/eui';
+import FieldSelect from './aggs/field_select';
+
+export function DataConfig({ onChange, model, fields }) {
+  const handleSelectChange = createSelectHandler(onChange);
+  const handleTextChange = createTextHandler(onChange);
```

Comment:
This totally works, so feel free to dismiss the following as too much unnecessary optimization. I just thought it would be a good opportunity to point out some performance best practices :wink: 

These change handlers are re-created on every render pass with new identities causing React to re-assign them on every reconciliation. Two good alternatives come to mind:

1. Create specialized stateful `MetricEditorInput` and `MetricEditorSelect` components that implement the logic of `createTextHandler` and `createSelectHandler` internally by wrapping the passed-in `onChange` prop. Then the usage would just be `<MetricEditorInput onChange=... />` and react would keep the component identities stable.
2. Create a wrapper component that keeps the function identities stable and passes the wrapped handler as arguments to its callable child prop:
```javascript
<TextChangeHandler name="index_pattern" defaultValue="someDefault" onChange={onChange}>
  {(wrappedOnChange) => (
    <input
      onChange={wrappedOnChange}
      ...
      />
  )}
</TextChangeHandler>
```

The second option would be more generic by virtue of not being bound to the actual rendered component.

Reply:
What about changing createSelectHandler to something like this...

```javascript
import _ from 'lodash';
const handlers = new WeakMap();
export default (handleChange) => {

  const handler = handlers.get(handleChange) || handleChange;
  if (!handlers.has(handler)) handlers.set(handleChange, handleChange);

  return (name) =>  {
    if (!handler._valueHandlers) handler._valueHandlers = {};
    if (handler._valueHandlers[name]) {
      console.log('Found previous value handler for ', name);
      return handler._valueHandlers[name];
    }

    const valueHandler = value => {
      if (_.isFunction(handleChange)) {
        return handleChange({
          [name]: value && value.value || null
        });
      }
    };

    handler._valueHandlers[name] = valueHandler;
    return valueHandler;
  };
};
```

If the onChange handler is the same and the name is the same then it just reuses the same function otherwise it creates a new function. This will allow us to keep the API exactly the same and fix everywhere `createSelectHandler` is used.

Reply:
Yes, memoization works too. :+1: [Recompose](https://github.com/acdlite/recompose/) has a generic implementation of this with [`withHandlers`](https://github.com/acdlite/recompose/blob/master/docs/API.md#withhandlers). The two solutions I mentioned have the same effect, just using React's virtual DOM mechanisms.

I'm fine with deferring this until it becomes a problem.

Reply:
Let's defer